File: day15_part2_2021.py

import heapq

class Position:
    def __init__(self, x, y, risk):
        self.x = x
        self.y = y
        self.risk = risk

    def __lt__(self, other):
        return self.risk < other.risk

def dijkstra(grid):
    pq = []
    heapq.heappush(pq, Position(0, 0, 0))

    rows = len(grid)
    cols = len(grid[0])
    dist = [[1 << 31 - 1 for _ in range(cols)] for _ in range(rows)]
    dist[0][0] = 0

    directions = [Position(1, 0, 0), Position(0, 1, 0), Position(-1, 0, 0), Position(0, -1, 0)]

    while pq:
        curr = heapq.heappop(pq)
        if curr.x == rows - 1 and curr.y == cols - 1:
            return curr.risk
        for d in directions:
            nx, ny = curr.x + d.x, curr.y + d.y
            if 0 <= nx < rows and 0 <= ny < cols:
                next_risk = curr.risk + grid[nx][ny]
                if next_risk < dist[nx][ny]:
                    dist[nx][ny] = next_risk
                    heapq.heappush(pq, Position(nx, ny, next_risk))

    return -1

def extend_grid(initial_grid):
    rows = len(initial_grid)
    cols = len(initial_grid[0])
    extended_grid = [[0 for _ in range(cols * 5)] for _ in range(rows * 5)]
    for i in range(rows * 5):
        for j in range(cols * 5):
            new_risk = initial_grid[i % rows][j % cols] + i // rows + j // cols
            if new_risk > 9:
                new_risk -= 9
            extended_grid[i][j] = new_risk
    return extended_grid

if __name__ == "__main__":
    with open("input.txt", "r") as file:
        initial_grid = [list(map(int, line.strip())) for line in file]

    extended_grid = extend_grid(initial_grid)
    print(dijkstra(extended_grid))

File: day15_part2_2021.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <queue>

struct Position {
    int x, y, risk;
};

struct CompareRisk {
    bool operator()(const Position& a, const Position& b) {
        return a.risk > b.risk;
    }
};

std::vector<std::vector<int> > extendGrid(std::vector<std::vector<int> >& initialGrid) {
    int rows = initialGrid.size();
    int cols = initialGrid[0].size();
    std::vector<std::vector<int> > extendedGrid(rows * 5, std::vector<int>(cols * 5, 0));

    for (int i = 0; i < rows * 5; i++) {
        for (int j = 0; j < cols * 5; j++) {
            int newRisk = initialGrid[i % rows][j % cols] + i / rows + j / cols;
            if (newRisk > 9) {
                newRisk -= 9;
            }
            extendedGrid[i][j] = newRisk;
        }
    }

    return extendedGrid;
}

int dijkstra(std::vector<std::vector<int> >& grid) {
    std::priority_queue<Position, std::vector<Position>, CompareRisk> pq;
    pq.push({0, 0, 0});

    int rows = grid.size();
    int cols = grid[0].size();
    std::vector<std::vector<int> > dist(rows, std::vector<int>(cols, INT_MAX));
    dist[0][0] = 0;

    std::vector<Position> directions = {{1, 0, 0}, {0, 1, 0}, {-1, 0, 0}, {0, -1, 0}};

    while (!pq.empty()) {
        Position curr = pq.top();
        pq.pop();

        if (curr.x == rows - 1 && curr.y == cols - 1) {
            return curr.risk;
        }

        for (const auto& d : directions) {
            int nx = curr.x + d.x;
            int ny = curr.y + d.y;

            if (nx >= 0 && ny >= 0 && nx < rows && ny < cols) {
                int nextRisk = curr.risk + grid[nx][ny];
                if (nextRisk < dist[nx][ny]) {
                    dist[nx][ny] = nextRisk;
                    pq.push({nx, ny, nextRisk});
                }
            }
        }
    }

    return -1;
}

int main() {
    std::ifstream file("input.txt");
    if (!file.is_open()) {
        std::cout << "Error opening file" << std::endl;
        return 1;
    }

    std::vector<std::vector<int> > initialGrid;
    std::string line;
    while (std::getline(file, line)) {
        std::vector<int> row;
        for (char ch : line) {
            row.push_back(ch - '0');
        }
        initialGrid.push_back(row);
    }

    file.close();

    std::vector<std::vector<int> > extendedGrid = extendGrid(initialGrid);

    std::cout << dijkstra(extendedGrid) << std::endl;

    return 0;
}
